samza-test/src/main/java/org/apache/samza/test/framework/system/CollectionStreamSystem.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.test.framework.system;
+
+import com.google.common.base.Preconditions;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.samza.system.inmemory.InMemorySystemFactory;
+
+
+/**
+ * CollectionStreamSystem represents a system that interacts with an underlying {@link InMemorySystemFactory} to create
+ * various input and output streams and initialize {@link org.apache.samza.system.SystemStreamPartition} with messages
+ * <p>
+ * Following system level configs are set by default
+ * <ol>
+ *   <li>"systems.%s.default.stream.samza.offset.default" = "oldest"</li>
+ *   <li>"systems.%s.samza.factory" = {@link InMemorySystemFactory}</li>
+ * </ol>
+ *
+ */
+public class CollectionStreamSystem {
+  private static final String SYSTEM_FACTORY = "systems.%s.samza.factory";
+  private static final String SYSTEM_OFFSET = "systems.%s.default.stream.samza.offset.default";
+
+  private String systemName;
+  private Map<String, String> systemConfigs;
+  private InMemorySystemFactory factory;
+
+  /**
+   * Constructs a new CollectionStreamSystem from specified components.
+   * <p>
+   * Every {@link CollectionStreamSystem} is assumed to consume from the oldest offset, since stream is in memory and
+   * is used for testing purpose. System uses {@link InMemorySystemFactory} to initialize in memory streams.
+   * <p>
+   * @param systemName represents unique name of the system
+   */
+  private CollectionStreamSystem(String systemName) {
```

Comment:
Could this be called InMemorySystem? It doesn't really seem like it has anything specific to "CollectionStream".

Reply:
I discussed the naming of this with Prateek and Xinyu, for the end users it gives them a better experience with future high level api implementation, hence named it CollectionStreamSystem

Reply:
I think the coupling makes sense to me now. I think what I was missing before was that TestRunner was coupling CollectionStream and CollectionStreamSystem by default since those were the only implementations of "stream" and "system". But it seems like if, in the future, there are other "streams" and "systems", then there will be some other way to more explicitly couple these.